dbms/src/Interpreters/JoinV2/HashJoinRowLayout.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Storages/KVStore/Utils.h>
+
+#include <vector>
+
+namespace DB
+{
+
+constexpr size_t ROW_ALIGN = 4;
+
+using RowPtr = char *;
+using RowPtrs = PaddedPODArray<RowPtr>;
+
+struct RowContainer
+{
+    PaddedPODArray<char> data;
+    PaddedPODArray<size_t> offsets;
+    PaddedPODArray<size_t> hashes;
+
+    size_t size() const { return offsets.size(); }
+
+    RowPtr getRowPtr(ssize_t row) { return &data[offsets[row - 1]]; }
+    size_t getHash(ssize_t row) { return hashes[row]; }
+};
+
+struct alignas(CPU_CACHE_LINE_SIZE) MultipleRowContainer
+{
+    std::mutex mu;
+    std::vector<RowContainer> column_rows;
+    size_t all_row_count = 0;
+
+    size_t build_table_index = 0;
+    size_t scan_table_index = 0;
+
+    void insert(RowContainer && row_container, size_t count)
+    {
+        std::unique_lock lock(mu);
+        column_rows.push_back(std::move(row_container));
+        all_row_count += count;
+    }
+
+    RowContainer * getNext()
+    {
+        std::unique_lock lock(mu);
+        if (build_table_index >= column_rows.size())
+            return nullptr;
+        return &column_rows[build_table_index++];
+    }
+
+    RowContainer * getScanNext()
```

Comment:
useless?

Reply:
It will be used for right outer/semi/anti join.